### PR TITLE
Allow GlobalWindows to be encoded as IntervalWindows

### LINKED ISSUE #25014: TypeError: Cannot convert GlobalWindow to apache_beam.utils.windowed_value._IntervalWindowBase
Hi Team,

We recently went into this error: `TypeError: Cannot convert GlobalWindow to apache_beam.utils.windowed_value._IntervalWindowBase`

Full error log:

`Error message from worker: generic::unknown: Traceback (most recent call last):
  File "apache_beam/runners/common.py", line 1458, in apache_beam.runners.common.DoFnRunner._invoke_bundle_method
  File "apache_beam/runners/common.py", line 561, in apache_beam.runners.common.DoFnInvoker.invoke_finish_bundle
  File "apache_beam/runners/common.py", line 566, in apache_beam.runners.common.DoFnInvoker.invoke_finish_bundle
  File "apache_beam/runners/common.py", line 1747, in apache_beam.runners.common._OutputHandler.finish_bundle_outputs
  File "apache_beam/runners/worker/operations.py", line 237, in apache_beam.runners.worker.operations.SingletonElementConsumerSet.receive
  File "apache_beam/runners/worker/operations.py", line 239, in apache_beam.runners.worker.operations.SingletonElementConsumerSet.receive
  File "apache_beam/runners/worker/operations.py", line 198, in apache_beam.runners.worker.operations.ConsumerSet.update_counters_start
  File "apache_beam/runners/worker/opcounters.py", line 215, in apache_beam.runners.worker.opcounters.OperationCounters.update_from
  File "apache_beam/runners/worker/opcounters.py", line 267, in apache_beam.runners.worker.opcounters.OperationCounters.do_sample
  File "apache_beam/coders/coder_impl.py", line 1472, in apache_beam.coders.coder_impl.WindowedValueCoderImpl.get_estimated_size_and_observables
  File "apache_beam/coders/coder_impl.py", line 1490, in apache_beam.coders.coder_impl.WindowedValueCoderImpl.get_estimated_size_and_observables
  File "apache_beam/coders/coder_impl.py", line 1222, in apache_beam.coders.coder_impl.SequenceCoderImpl.estimate_size
  File "apache_beam/coders/coder_impl.py", line 1238, in apache_beam.coders.coder_impl.SequenceCoderImpl.get_estimated_size_and_observables
  File "apache_beam/coders/coder_impl.py", line 208, in apache_beam.coders.coder_impl.CoderImpl.get_estimated_size_and_observables
  File "apache_beam/coders/coder_impl.py", line 829, in apache_beam.coders.coder_impl.IntervalWindowCoderImpl.estimate_size
TypeError: Cannot convert GlobalWindow to apache_beam.utils.windowed_value._IntervalWindowBase`

our pipeline looks something like this:
`          pipeline
            | "Read PubSub Messages"
            >> beam.io.ReadFromPubSub(
                subscription=options.input_subscription, with_attributes=False
            )
            | f"Group Messages into {options.window_size} seconds window"
            >> beam.WindowInto(
                window.FixedWindows(options.window_size),
                allowed_lateness=Duration(seconds=1 * 60 * 60),
            )  # 1 hour
            | "Extract data from requests" >> beam.ParDo(ExtractDataFromRequest()
`

This pipeline has been running successfully for a couple months and these errors only shown up recently and it seems nowhere to start debugging with. Any ideas?

### PR DESCRIPTION
**Pull request by damccorm**
Right now, we window outputs from bigquery write connector (representing failures) back into the global window, but we do this as part of the transform instead of in an explicit windowing step. This fails because when we do sampling, it expects the value to be in the window which was passed in. This behavior exists in other places and is supported in other SDKs, but doesn't work in python because we aren't able to encode the window correctly. This fixes the problem and allows the encoding to be a bit more permissive.

Without my change, this fails when performing sampling (example on [Dataflow](https://console.cloud.google.com/dataflow/jobs/us-central1/2024-09-26_09_37_14-7677325219113638771;graphView=0?project=apache-beam-testing))

After my change, this succeeds.  I also added a representative test

Fixes #25014

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by damccorm** [comment]
R: @Abacn

**Comment by damccorm** [comment]
R: @Abacn

**Comment by damccorm** [comment]
I'll note that there's probably a better fix here to introspect the function and figure out if it can return globally windowed things ahead of time so we can get our coders right. That's also a larger more ambiguous problem, and I'm hoping we can just get BQ IO (and other IOs which use this pattern like spanner) right for now

File: sdks/python/apache_beam/coders/coder_impl.py [hunk 1]
```diff
@@ -822,15 +824,19 @@ def _from_normal_time(self, value):
 
   def encode_to_stream(self, value, out, nested):
     # type: (IntervalWindow, create_OutputStream, bool) -> None
-    typed_value = value
+    if not TYPE_CHECKING:
```
Comment (robertwb) [inline sdks/python/apache_beam/coders/coder_impl.py:827]: We should not be checking this at every element encoding; instead lift the import.

File: sdks/python/apache_beam/utils/windowed_value_test.py [hunk 2]
```diff
@@ -75,6 +76,22 @@ def test_pickle(self):
...
 
+  def test_encoding_global_window_in_interval_window(self):
+    input_data = ['123']
+
+    class ComputeWordLengthFn(beam.DoFn):
+      def process(self, element):
+        pass
+
+      def finish_bundle(self):
+        yield beam.transforms.window.GlobalWindows.windowed_value('test')
```
Comment (robertwb) [inline sdks/python/apache_beam/utils/windowed_value_test.py:87]: One should only emit to windows that were part of the input.

**Comment by damccorm** [comment]
Superceded by https://github.com/apache/beam/pull/32583

**Pull request closed without merging**